slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -46,64 +88,140 @@ SLAMServiceActionMode SLAMServiceImpl::GetActionMode() {
+            map_builder.OverwriteMaxSubmapsToKeep(max_submaps_to_keep);
+        }
+        if (slam_action_mode == SLAMServiceActionMode::UPDATING) {
+            map_builder.OverwriteFreshSubmapsCount(fresh_submaps_count);
+            map_builder.OverwriteMinCoveredArea(min_covered_area);
+            map_builder.OverwriteMinAddedSubmapsCount(min_added_submaps_count);
+        }
+        map_builder.OverwriteOccupiedSpaceWeight(occupied_space_weight);
+        map_builder.OverwriteTranslationWeight(translation_weight);
+        map_builder.OverwriteRotationWeight(rotation_weight);
     }
-
-    auto mutable_ceres_scan_matcher_options =
-        mutable_pose_graph_options->mutable_constraint_builder_options()
-            ->mutable_ceres_scan_matcher_options();
-    mutable_ceres_scan_matcher_options->set_occupied_space_weight(
-        occupied_space_weight);
-    mutable_ceres_scan_matcher_options->set_translation_weight(
-        translation_weight);
-    mutable_ceres_scan_matcher_options->set_rotation_weight(rotation_weight);
 }
 
 void SLAMServiceImpl::SetUpMapBuilder() {
     auto action_mode = GetActionMode();
     if (action_mode == SLAMServiceActionMode::MAPPING) {
-        mapBuilder.SetUp(this->configuration_directory,
-                         this->configuration_mapping_basename);
+        {
+            std::lock_guard<std::mutex> lk(map_builder_mutex);
+            map_builder.SetUp(this->configuration_directory,
+                              this->configuration_mapping_basename);
+        }
     } else if (action_mode == SLAMServiceActionMode::LOCALIZING) {
-        mapBuilder.SetUp(this->configuration_directory,
-                         this->configuration_localization_basename);
+        {
+            std::lock_guard<std::mutex> lk(map_builder_mutex);
+            map_builder.SetUp(this->configuration_directory,
+                              this->configuration_localization_basename);
+        }
     } else if (action_mode == SLAMServiceActionMode::UPDATING) {
-        mapBuilder.SetUp(this->configuration_directory,
-                         this->configuration_update_basename);
+        {
+            std::lock_guard<std::mutex> lk(map_builder_mutex);
+            map_builder.SetUp(this->configuration_directory,
+                              this->configuration_update_basename);
+        }
     } else {
         throw std::runtime_error("invalid action mode");
     }
     OverwriteMapBuilderParameters();
-    mapBuilder.BuildMapBuilder();
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
+        map_builder.BuildMapBuilder();
+    }
+}
+
+void SLAMServiceImpl::PaintMap(std::string &jpeg_img) {
+    const double kPixelSize = 0.01;
+    cartographer::mapping::MapById<
+        cartographer::mapping::SubmapId,
+        cartographer::mapping::PoseGraphInterface::SubmapPose>
+        submap_poses;
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
```

Comment:
If possible, I think we should lock map_builder only once and get everything we need from it, so that we get a consistent map.

Reply:
I understand the concern and share it. The problem with `PaintMap` is that it takes a very long time to create the map: Between 2 - 30 seconds (see this ticket:https://viam.atlassian.net/browse/DATA-661). Locking `map_builder` for the entire time would mean mapping would be stopped for this long, which we can't do.

I investigated whether or not the map might change **while** we're painting it, and from what I understand and what I've seen in the practical tests, I believe that the map won't change during the printing process (dramatically). 

Given that currently map painting takes too long in order for us to block `map_builder` the entire time, and that we have a ticket for investigating if we can reduce this time, I have created another ticket that'll investigate whether or not we should block `map_builder` during the entire time while painting the map - or if we can find a way to extract all necessary information at once and THEN paint the map: https://viam.atlassian.net/browse/DATA-668

Reply:
Ah, right, this became an issue when I am calling `slamService.map_builder...` in the `slam_service_test.cc`, like this:

```cpp
void checkCartoMapBuilderParameters(SLAMServiceImpl& slamService) {
    auto pose_graph_options =
        slamService.map_builder.map_builder_options_.pose_graph_options();
...
```

resulting in this error:

```bash
/Users/kat/slam/slam-libraries/viam-cartographer/src/slam_service/slam_service_test.cc:13:21: error: 'map_builder' is a private member of 'viam::SLAMServiceImpl'
        slamService.map_builder.map_builder_options_.pose_graph_options();
```

Do you have suggestions on how I can resolve this elegantly? 

Something that comes to my mind is that I could implement functions within the `slamService` that retrieve & return those parameters from `map_builder`. Would this be the way to go here? Or are there alternative solutions?

Reply:
That is a long time! I agree we'll need to investigate it.

When I wrote my comment, I was imagining having this ticket do the work of extracting the necessary information from `map_builder` under the lock and then paint the map. It looks like that's what you described in DATA-668. The reason I thought we should do it now is because the behavior may not be well-defined when accessing an inconsistent view of the map. For example, the set of submap poses could change between when we get their IDs on line 142 and when we call `SubmapToProto()` on line 155 (though you do catch any errors in case an ID no longer exists). But if you've found that in practice, the map has changed little, and we're usually able to get a map without error that looks correct, then maybe it's fine to defer this investigation until DATA-668.

In the long-term (perhaps this is for DATA-661), I wonder if we should use cartographer's map-painting capability at all. `GetMap()` will need to be able to return a pointcloud as well, so we could transform the pointcloud into a jpeg in the same way that orbslam does. We would lose out on cartographer's pretty maps, but maybe that would be faster. And in the even longer-term, we won't be making jpegs within the slam libraries at all, since they will always return octrees, and the octrees will get converted to jpegs in rdk, so maybe we shouldn't spend too much time worrying about this.

Reply:
In answer to your second comment, yes, you could create getters for each of the `map_builder` parameters. A couple other options are:
- Add a function like `CheckMapBuilderParameters()` that takes in the expected parameters and compares them to the actual `map_builder` parameters
- Add a function `GetMapBuilderForTesting()` that returns the underlying `map_builder`. This should only be used in tests and the test is responsible for any concurrency requirements

Reply:
I agree with the idea of not using cartographer's map-painting capability - especially due to the color issues it causes. I believe we could implement our own version (similar to what you suggest with how orbslam implements it, though I would still do it similarly to how PaintMap does it, rather than through the point cloud, since important information would be lost that way). I do think we could speed map painting up that way and solve the color issues at the same time. It's good you mention the longer-term plan of using octrees to create a jpeg - which does really mean it's not worth putting too much effort into this.

Let me escalate the ticket that'll solve the issue we're discussing here to one of the top ones, so this will be fixed in a timely manner.